File: solutions/csharp/clock/1/Clock.cs
using System;

public class Clock
{
    private readonly int hours;
    private readonly int minutes;

    public Clock(int hours, int minutes)
        : this(minutes + (hours * 60))
    {
    }

    private Clock(int totalMinutes)
    {
        (this.hours, this.minutes) = GetHoursAndMinutes(totalMinutes);
    }

    public Clock Add(int minutesToAdd)
    {
        return new Clock(minutes + (hours * 60) + minutesToAdd);
    }

    public Clock Subtract(int minutesToSubtract)
    {
        return new Clock(minutes + (hours * 60) - minutesToSubtract);
    }

    public override string ToString()
    {
        return $"{hours:00}:{minutes:00}";
    }

    public override bool Equals(object obj)
    {
        var other = obj as Clock;
        if (other == null)
        {
            return false;
        }

        return hours == other.hours && minutes == other.minutes;
    }

    private (int hours, int minutes) GetHoursAndMinutes(int totalMinutes)
    {
        var minutes = (totalMinutes % 60 + (totalMinutes < 0 ? 60 : 0)) % 60;

        var totalHours = (totalMinutes - minutes) / 60;
        var hours = (totalHours % 24 + (totalHours < 0 ? 24 : 0)) % 24;
        return (hours, minutes);
    }
}

File: solutions/go/clock/1/clock.go
// Package clock implements a simple library that implements
// methods to handle times without dates
package clock

import (
	"fmt"
)

// Clock struct to keep time in minutes
type Clock struct {
	timeInMinutes int
}

// New returns a new instance of Clock
func New(hours, minutes int) Clock {
	maxMinutes := 24 * 60
	totalMinutes := hours*60 + minutes
	return Clock{
		timeInMinutes: (maxMinutes + totalMinutes%maxMinutes) % maxMinutes,
	}
}

// Add method adds minutes to a given Clock and returns a new instance
func (c Clock) Add(minutesToAdd int) Clock {
	return New(c.hours(), c.minutes()+minutesToAdd)
}

// Subtract method subtracts minutes from a given Clock and returns a new instance
func (c Clock) Subtract(minutesToSubtract int) Clock {
	return c.Add(-minutesToSubtract)
}

// Valid returns square of sum of first n natural numbers, if successful
func (c Clock) String() string {
	return fmt.Sprintf("%02d:%02d", c.hours(), c.minutes())
}

func (c Clock) hours() int {
	return c.timeInMinutes / 60
}

func (c Clock) minutes() int {
	return c.timeInMinutes % 60
}
